server/internal/server/models.go

```diff
@@ -208,75 +206,112 @@ func (s *S) ListModels(
 		return nil, status.Errorf(codes.Internal, "get total items: %s", err)
 	}
 
-	var modelProtos []*v1.Model
-
-	if req.After == "" || afterBaseModel != nil {
-		// First include base models.
-		var afterModelID string
-		if afterBaseModel != nil {
-			afterModelID = afterBaseModel.ModelID
-		}
-
-		bms, hasMore, err := s.store.ListBaseModelsWithPagination(userInfo.TenantID, afterModelID, int(limit), req.IncludeLoadingModels)
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "list base models: %s", err)
-		}
-		for _, m := range bms {
-			as, err := getModelActivationStatus(s.store, m.ModelID, m.TenantID)
+	startCat := 0
+	afterID := ""
+	if req.After != "" {
+		var as v1.ActivationStatus
+		var isBase bool
+		if afterBase != nil {
+			as, err = getModelActivationStatus(s.store, afterBase.ModelID, afterBase.TenantID)
 			if err != nil {
-				return nil, err
+				return nil, status.Errorf(codes.Internal, "get model activation status: %s", err)
 			}
-			mp, err := baseToModelProto(m, as)
+			isBase = true
+		} else {
+			as, err = getModelActivationStatus(s.store, afterModel.ModelID, afterModel.TenantID)
 			if err != nil {
-				return nil, status.Errorf(codes.Internal, "to proto: %s", err)
+				return nil, status.Errorf(codes.Internal, "get model activation status: %s", err)
 			}
-			modelProtos = append(modelProtos, mp)
+			isBase = false
 		}
+		startCat = activationCategory(isBase, as)
```

Comment:
`activationCategory` and `categories` defined below has some overlap. 

my preference is 

- define `categories` at the top of the function
- define `activationCategory` as a local function that uses `categories`

```go
activationCategory = func(isBase bool, status v1.ActivationStatus) int {
  for i, cat := range categories {
     if cat.isBase == isBase &7 cat.status == status {
      return i
    }
  }
  return -1 // or return error
}